pkg/catalog/catalog.go

```diff
@@ -1748,67 +1753,213 @@ func (c *Catalog) FindMergeBase(ctx context.Context, repositoryID string, destin
 }
 
-func (c *Catalog) DumpCommits(ctx context.Context, repositoryID string) (string, error) {
+func (c *Catalog) DumpRefsSubmit(ctx context.Context, repositoryID string) (string, error) {
 	repository, err := c.getRepository(ctx, repositoryID)
 	if err != nil {
 		return "", err
 	}
 
-	metaRangeID, err := c.Store.DumpCommits(ctx, repository)
-	if err != nil {
+	// create refs dump task and update initial status
+	taskID := newTaskID(DumpRefsTaskIDPrefix)
+	taskStatus := &RefsDumpStatus{
+		Task: &Task{
+			Id:        taskID,
+			Completed: false,
+			UpdatedAt: timestamppb.Now(),
+		},
+		Info: &RefsDumpInfo{},
+	}
+	if err := updateTaskStatus(ctx, c.KVStore, repository, taskID, taskStatus); err != nil {
 		return "", err
 	}
-	return string(*metaRangeID), nil
+	c.workPool.Submit(func() {
+		c.dumpRefsTask(ctx, err, repository, taskStatus, taskID)
+	})
+	return taskID, nil
+}
+
+func (c *Catalog) dumpRefsTask(ctx context.Context, err error, repository *graveler.RepositoryRecord, taskStatus *RefsDumpStatus, taskID string) {
+	steps := []struct {
+		name string
+		fn   func() error
+	}{
+		{
+			name: "dump commits",
+			fn: func() error {
+				var commitsMetaRangeID *graveler.MetaRangeID
+				commitsMetaRangeID, err = c.Store.DumpCommits(ctx, repository)
+				if err != nil {
+					return err
+				}
+				taskStatus.Info.CommitsMetarangeId = string(*commitsMetaRangeID)
+				return nil
+			},
+		},
+		{
+			name: "dump branches",
+			fn: func() error {
+				var branchesMetaRangeID *graveler.MetaRangeID
+				branchesMetaRangeID, err = c.Store.DumpBranches(ctx, repository)
+				if err != nil {
+					return err
+				}
+				taskStatus.Info.BranchesMetarangeId = string(*branchesMetaRangeID)
+				return nil
+			},
+		},
+		{
+			name: "dump tags",
+			fn: func() error {
+				var tagsMetaRangeID *graveler.MetaRangeID
+				tagsMetaRangeID, err = c.Store.DumpTags(ctx, repository)
+				if err != nil {
+					return err
+				}
+				taskStatus.Info.TagsMetarangeId = string(*tagsMetaRangeID)
+				return nil
+			},
+		},
+	}
+
+	// run each step and update step status
+	for stepIdx, step := range steps {
+		err := step.fn()
+		if err != nil {
+			c.log(ctx).
+				WithError(err).
+				WithFields(logging.Fields{"task_id": taskID, "repository": repository.RepositoryID}).
+				Errorf("failed to %s", step.name)
+			taskStatus.Task.Completed = true
+			taskStatus.Task.Error = err.Error()
+		} else if stepIdx == len(steps)-1 {
+			taskStatus.Task.Completed = true
+		}
+		taskStatus.Task.UpdatedAt = timestamppb.Now()
+
+		err = updateTaskStatus(ctx, c.KVStore, repository, taskID, taskStatus)
+		if err != nil {
+			c.log(ctx).
+				WithError(err).
+				WithFields(logging.Fields{"task_id": taskID, "repository": repository.RepositoryID}).
+				Error("failed to update task status")
+		}
+
+		// make sure we stop based on task completed status, as we may fail
+		if taskStatus.Task.Completed {
+			break
+		}
+	}
```

Comment:
This is duplicate code - can we extract it?

Reply:
will use backoff common code to run the loop - but each one perform different call with different output type